SpatialGDK/Source/SpatialGDK/Private/Interop/Startup/AssignedPartitionStartupSteps.cpp

```diff
@@ -0,0 +1,144 @@
+#include "Interop/GlobalStateManager.h"
+#include "SpatialView/EntityComponentTypes.h"
+
+namespace SpatialGDK
+{
+class FReadVirtualWorkerTranslatorStep : public FStartupStep
+{
+public:
+	FReadVirtualWorkerTranslatorStep(TSharedRef<FInternalState> InState, TSharedRef<TOptional<Worker_PartitionId>> InLocalPartitionIdPtr,
+									 ISpatialOSWorker& InConnection, UGlobalStateManager& InGlobalStateManager);
+
+	virtual bool TryFinish() override;
+
+private:
+	TSharedRef<const FInternalState> State;
+	TSharedRef<TOptional<Worker_PartitionId>> LocalPartitionIdPtr;
+
+	ISpatialOSWorker* Connection;
+	UGlobalStateManager* GlobalStateManager;
+};
+
+class FWaitForAssignedPartition : public FStartupStep
+{
+public:
+	FWaitForAssignedPartition(TSharedRef<FInternalState> InState, TSharedRef<const TOptional<Worker_PartitionId>> InLocalPartitionIdPtr,
+							  USpatialNetDriver& InNetDriver, ISpatialOSWorker& InConnection);
+
+	virtual bool TryFinish() override;
+
+private:
+	TSharedRef<const FInternalState> State;
+	TSharedRef<const TOptional<Worker_PartitionId>> LocalPartitionIdPtr;
+
+	USpatialNetDriver* NetDriver;
+	ISpatialOSWorker* Connection;
+};
+
+FGetAssignedPartitionStep::FGetAssignedPartitionStep(TSharedRef<FInternalState> InState, USpatialNetDriver& InNetDriver)
+	: State(InState)
+	, NetDriver(&InNetDriver)
+	, LocalPartitionId(MakeShared<TOptional<Worker_PartitionId>>())
+	, Executor(CreateSteps())
+{
+}
+
+TArray<TUniquePtr<FStartupStep>> FGetAssignedPartitionStep::CreateSteps()
+{
+	TArray<TUniquePtr<FStartupStep>> Steps;
+
+	Steps.Emplace(MakeUnique<FReadVirtualWorkerTranslatorStep>(State, LocalPartitionId, NetDriver->Connection->GetCoordinator(),
+															   *NetDriver->GlobalStateManager));
+	Steps.Emplace(MakeUnique<FWaitForAssignedPartition>(State, LocalPartitionId, *NetDriver, NetDriver->Connection->GetCoordinator()));
+
+	return Steps;
+}
+
+FReadVirtualWorkerTranslatorStep::FReadVirtualWorkerTranslatorStep(TSharedRef<FInternalState> InState,
+																   TSharedRef<TOptional<Worker_PartitionId>> InLocalPartitionIdPtr,
+																   ISpatialOSWorker& InConnection,
+																   UGlobalStateManager& InGlobalStateManager)
+	: State(InState)
+	, LocalPartitionIdPtr(InLocalPartitionIdPtr)
+	, Connection(&InConnection)
+	, GlobalStateManager(&InGlobalStateManager)
+{
+	StepName = TEXT("Retrieving VirtualWorkerTranslator assigned partition");
+}
+
+bool FReadVirtualWorkerTranslatorStep::TryFinish()
+{
+	const ComponentData* DeploymentMapComponent =
+		Connection->GetView()[SpatialConstants::INITIAL_GLOBAL_STATE_MANAGER_ENTITY_ID].Components.FindByPredicate(
+			ComponentIdEquality{ SpatialConstants::DEPLOYMENT_MAP_COMPONENT_ID });
+	if (ensure(DeploymentMapComponent != nullptr))
+	{
+		const int32 ExistingSessionId = Schema_GetInt32(DeploymentMapComponent->GetFields(), SpatialConstants::DEPLOYMENT_MAP_SESSION_ID);
+
+		GlobalStateManager->ApplySessionId(ExistingSessionId);
```

Comment:
Is there a risk of something going wrong if this bit of the step is done several times? It looks like its own individual step, or something that could be done as part of a previous step that retrieved the GSM entity.

Reply:
Moved to a separate step.